armadillo/src/main/java/at/favre/lib/armadillo/SharedPreferenceChangeListenerWrapper.java

```diff
@@ -0,0 +1,56 @@
+package at.favre.lib.armadillo;
+
+import android.content.SharedPreferences;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.lang.ref.WeakReference;
+
+/**
+ * Implementation that will wrap a {@link OnSecurePreferenceChangeListener} to adapt to the standard {@link SharedPreferences.OnSharedPreferenceChangeListener}
+ * @since 03.07.2019
+ */
+final class SharedPreferenceChangeListenerWrapper implements SharedPreferences.OnSharedPreferenceChangeListener {
+
+    private static class KeyComparissonImpl implements OnSecurePreferenceChangeListener.DerivedKeyComparison {
+        private final EncryptionProtocol encryptionProtocol;
+
+        private KeyComparissonImpl(EncryptionProtocol encryptionProtocol) {
+            this.encryptionProtocol = encryptionProtocol;
+        }
+
+        @Override
+        public boolean isDerivedKeyEqualTo(@NonNull String derivedKey, @NonNull String key) {
+            return derivedKey.equals(encryptionProtocol.deriveContentKey(key));
+        }
+    }
+
+    private final OnSecurePreferenceChangeListener.DerivedKeyComparison keyComparison;
+    private final WeakReference<OnSecurePreferenceChangeListener> wrappedRef;
```

Comment:
I think using `WeakReference `here might lead to problems. This may be not the behaviour you want and it is quite difficult to understand from the outside why this may not work for a developer. I usually refrain from using Weak/Soft references as they not very clear to most devs how they work (and behave differently on Android ie. more aggresivly). I would make it the callers responsibility to unregister his listeners, as it is usually the case in Android.
As reference why this might be an issue: https://android.jlelse.eu/android-weakreferences-how-not-to-5522bb325935

Reply:
I know and I myself did suffer some issues at this respect letting Kotlin's lambda to implement directly a `OnSharedPreferencesChangeListener` which was translated to an anonymous class instantiation, thus being garbage collected (I didn't know at the time how kotlin translated that piece of code, I learnt it by force).

I decided to go with the `WeakReference` approach just based on the same principle than the `SharedPreferences`, because well the system indeed treats `SharedPreferences` instances as singletons, then adding listeners and failing to unregister could cause potential leak. 

For me is a design choice, if you prefer I can remove the WeakReferences, but here I like them indeed, for same reason than a user should know well to not register twice a listener, they should know well to keep a reference to the listener (I could add this fact to the javadoc). It's also true that armadillo isn't enforcing any singleton and will be the user of the library who decides how he/she manages that.

Reply:
Fair argument, however I prefer the easier to understand, explicit way with a little more boilerplate, ie. please change it to register/unregister with a strong reference.

Reply:
Ok, I've just updated to avoid usage of WeakReferences on listeners